Plutarch/DataRepr/Internal.hs

```diff
@@ -65,17 +91,63 @@ import Plutarch.Builtin (
 import Plutarch.DataRepr.Internal.Generic (MkSum (mkSum))
 import Plutarch.DataRepr.Internal.HList (type Drop, type IndexList)
 import Plutarch.Integer (PInteger)
-import Plutarch.Internal (S (SI))
-import Plutarch.Lift (PConstant, PConstantRepr, PConstanted, PLift, pconstantFromRepr, pconstantToRepr)
-import Plutarch.List (pdrop, ptryIndex)
+import Plutarch.Internal (S (SI), punsafeBuiltin)
+import Plutarch.Internal.TypeFamily (ToPType2)
+import Plutarch.Lift (PConstant, PConstantRepr, PConstanted, PLift, pconstant, pconstantFromRepr, pconstantToRepr)
+import Plutarch.List (PListLike (pnil), pcons, pdrop, ptryIndex)
 import Plutarch.TermCont (TermCont, hashOpenTerm, runTermCont)
 import Plutarch.Unsafe (punsafeCoerce)
 import qualified Plutus.V1.Ledger.Api as Ledger
+import qualified PlutusCore as PLC
 
 {- | A "record" of `exists a. PAsData a`. The underlying representation is
  `PBuiltinList PData`.
 -}
-data PDataRecord (as :: [PLabeledType]) (s :: S)
+data PDataRecord (as :: [PLabeledType]) (s :: S) where
+  PDCons ::
+    forall name x xs s.
+    Term s (PAsData x) ->
+    (Term s (PDataRecord xs)) ->
+    PDataRecord ((name ':= x) ': xs) s
+  PDNil :: PDataRecord '[] s
+
+instance PlutusType (PDataRecord ((name ':= x) ': xs)) where
+  type PInner (PDataRecord ((name ':= x) ': xs)) _ = PBuiltinList PData
+  pcon' (PDCons x xs) = punsafeCoerce result
+    where
+      result :: Term _ (PDataRecord ((name ':= x) ': xs))
+      result = pdcons # x # xs
+  pmatch' x f = pmatch x $ \case
+    PCons x' xs' ->
```

Comment:
We already know the list is non-empty. The reason for this is that on line 114, we declare the instance for only non-empty data records. You can just use `phead`.

Reply:
Yes, but we need the head and tail. `ChooseList` (what `pmatch` will expand to here) is going to more efficient than head + tail. At the cost of also having to handle the absurd nil case.

Reply:
I'm not entirely sure what you mean. `pmatch` also just gets the head and tail separately, it only uses `ChooseList` to figure out whether it's a head or tail IIRC. Look at your own PR! https://github.com/Plutonomicon/plutarch/pull/217/files

Reply:
Oh lol - I was under the impression choose list takes in lambdas. Never mind! Updated.